python/ray/dag/compiled_dag_node.py

```diff
@@ -1864,11 +1909,62 @@ def _check_inputs(self, args: Tuple[Any, ...], kwargs: Dict[str, Any]) -> None:
                     f"must be called with kwarg `{kwarg}`"
                 )
 
+    def _has_execution_results(
+        self,
+        execution_index: int,
+    ) -> bool:
+        """Check whether there are results corresponding to the given execution
+        index stored in self._result_buffer. This helps avoid fetching and
+        storing results again.
+
+        Args:
+            execution_index: The execution index corresponding to the result.
+
+        Returns:
+            Whether the result for the given index has been fetched and cached.
+        """
+        return execution_index in self._result_buffer
+
+    def _cache_execution_results(
+        self,
+        execution_index: int,
+        channel_index: Optional[int],
+        result: Optional[Any] = None,
+    ) -> List[Any]:
+        """Cache execution results in self._result_buffer and return the result. Results
+        are converted to dictionary format to allow efficient element removal and
+        calculation of buffer size.
+
+        Args:
+            execution_index: The execution index corresponding to the result.
+            channel_index: The index of the output channel corresponding to the result.
+                Channel indexing is consistent with the order of
+                self.dag_output_channels. None means that the result wraps outputs from
+                all output channels.
+            result: The result to be cached. None means the result has been cached.
+
+        Returns:
+            The execution result corresponding to the given execution index and channel
+            index.
+        """
+        if execution_index not in self._result_buffer:
+            for chan_idx, res in enumerate(result):
+
+                self._result_buffer[execution_index][chan_idx] = res
+
+        if channel_index is None:
+            del self._result_buffer[execution_index]
```

Comment:
Doesn't this need to set `result`? Please add a test if not tested already.

Reply:
Yes, this should. This could potentially return incorrect results when CompiledDAGFutures are fetched in order. Will add a test as well.